source/fundamentals/promises.txt

```diff
@@ -0,0 +1,147 @@
+``countDocuments()``, or ``update()``. Once the operation completes execution,
+the callback method executes.
+
+The callback method you provide should match the method signature of the 
+**type** specified in the :node-api:`API documentation <>`.
+
+ .. note::
+ 
+    If you specify a callback, the method *does not* return a Promise.
+
+Operational Considerations
+--------------------------
+
+When using ``async`` methods, make sure to use ``await`` when you call methods
+that return Promises to ensure they complete before executing any other
+logic that depends on the result. Consider the following example in which
+we iterate over a cursor using ``hasNext()``, which returns a Promise that
+resolves to a boolean that indicates whether additional results exist, and
+``next()`` which returns a Promise that resolves to the next entry the
+cursor is pointing to.
+
+One common mistake is to forget to use ``await`` to get the value of the
+result rather than the Promise object. Consider the following code:
+
+.. code-block:: js
+
+   // WARNING: this snippet may cause an infinite loop 
+   const cursor = collection.find();
+
+   while (cursor.hasNext()) {
+      console.log(cursor.next());
+   }
+
+Since the call to ``hasNext()`` returns a ``Promise``, the conditional
+statement returns ``true`` regardless of the value that it resolves to.
+
+If we alter the code to ``await`` the call to ``next()`` only, as demonstrated
+in the following code snippet, it throws the following error: 
+``MongoError: Cursor is closed``.
+
+.. code-block:: js
+   
+   // WARNING: this snippet throws a MongoError
+   const cursor = collection.find();
+
+   while (cursor.hasNext()) {
+      console.log(await cursor.next());
+   }
+
+While ``hasNext()`` is not called until after the result of ``next()`` returns,
+the call to ``hasNext()`` returns a Promise which evaluates to ``true`` rather
+than the value it resolves to, similar to the prior example. The code
+attempts to call ``next()`` on a Cursor that has already returned its results
+and closed as a result.
+
+If we alter the code to only ``await`` the call to ``hasNext()`` as shown in
+the following example, the console prints Promise objects rather than the
+document objects. 
+
+.. code-block:: js
+
+   // WARNING: this snippet prints Promises instead of the objects they resolve to
+   const cursor = collection.find();
+
+   while (await cursor.hasNext()) {
+      console.log(cursor.next());
+   }
+
+Use ``await`` before both the ``hasNext()`` and ``next()`` method calls to
+ensure that you are operating on the correct return values as demonstrated
+in the following code:
+
+.. code-block:: js
+
+   const cursor = collection.find();
+
+   while (await cursor.hasNext()) {
+      console.log(await cursor.next());
+   } 
+
+.. note::
+
+   For additional information on using Promises and Callbacks with the MongoDB
+   NodeJS driver, see this
+   `MongoDB University m220js course video <https://youtu.be/R4AEyKehpss>`_.
```

Comment:
it might be helpful to change the link text to:  see this Mongodb University M220js course video on `asynchronous programming <youtube link>`_. So that the user gets more of a preview on what they're clicking (a video about async programming) rather than the general  term "MongoDB University m220js course video"